### PR TITLE
Add OpenBSD specific build targets

### PR DESCRIPTION
**Pull request by devnexen**
-Wl,-s,defs causing linkage issue stumbled accross somewhat
similar issues as #11043.

### CODE REVIEW

**Comment by levitte** [comment]
"somewhat similar", although I assume you got issues building engines.  This PR is a "quick fix", but it drops our attempt to make engines and other DSOs self-contained.  I would very much like to see a log of what went wrong before accepting this fix, it's quite possible that you have uncovered a build issue that should be addressed in a different (better?) manner.

**Comment by devnexen** [comment]
> "somewhat similar", although I assume you got issues building engines. This PR is a "quick fix", but it drops our attempt to make engines and other DSOs self-contained. I would very much like to see a log of what went wrong before accepting this fix, it's quite possible that you have uncovered a build issue that should be addressed in a different (better?) manner.

I get pipe, close, write ... as undefined on OpenBSD, it seems this linker flag is broken in this platform.

**Comment by levitte** [comment]
> I get pipe, close, write ... as undefined on OpenBSD, it seems this linker flag is broken in this platform.

???!!!

Wow...  yeahok, if it's at that level, I can't see a better solution.

**Comment by levitte** [comment]
Is this for 1.1.1 or master or both?

**Review by levitte — APPROVED** [review]

**Comment by devnexen** [comment]
Well master for now I might say.

**Comment by levitte** [comment]
> Well master for now I might say.

Thanks

**Review by paulidale — APPROVED** [review]

**Comment by kaduk** [comment]
What linker is OpenBSD using?
I am worried that this might introduce regressions for other BSDs (e.g., FreeBSD).

**Comment by devnexen** [comment]
The LLVM's

**Comment by kaduk** [comment]
> I get pipe, close, write ... as undefined on OpenBSD, it seems this linker flag is broken in this platform.

Interesting!  The "undefined symbol" does seem to match up with #11043 (though I would always prefer to see the actual diagnostic output that you are seeing), but does not match what I've been able to see so far on FreeBSD.
Since the sanitizers are implicated in the (non-)inclusion of `-z defs` that are affected by this proposed change, I tried building on FreeBSD configured with and without `enable-asan`.  The normal build went fine, but the `enable-asan` build dies trying to link `fuzz/client-test`, complaining about a *duplicate* symbol (time), defined both in /usr/lib/clang/11.0.0/lib/freebsd/libclang_rt.asan-x86_64.a and fuzz/client.c:38 .  Most interestingly, `defs` does not appear in either build log!

**Comment by devnexen** [comment]
Maybe, since it affects all BSD, the solution is to close it and allows OpenBSD to locally patch their OpenSSL port.

**Comment by kaduk** [comment]
> Maybe, since it affects all BSD, the solution is to close it and allows OpenBSD to locally patch their OpenSSL port.

I don't think we need to resort to that quite yet.  For one, my testing so far has just been on unpatched openssl sources -- I should also try it after applying this PR (or you could, if you're in a position to do so faster than me).

**Comment by devnexen** [comment]
same outcome with this PR.

**Comment by kaduk** [comment]
> Interesting! The "undefined symbol" does seem to match up with #11043 (though I would always prefer to see the actual diagnostic output that you are seeing), but does not match what I've been able to see so far on FreeBSD.

Oops, it seems that my initial testing was accidentally based on a 1.1.1 tree and not the master-based tree I thought I was using.  I will need to run them again...

**Comment by kaduk** [comment]
Okay, my test results make more sense now that I am testing what I think I am testing :smile: 

I would still very much like to see the build log for the actual failures that this change is fixing; the stuff linked in #11043 is different enough that it's hard to draw many conclusions from it.

Assuming that seeing the build log doesn't bring further enlightenment, I'm inclined to think that the "right thing to do" is to split off a dedicated openbsd configuration stanza to override the dso_ldflags and teach our `config` how to guess it as the target to pass to `Configure`.

**Comment by devnexen** [comment]
`
ld: error: undefined symbol: strcmp
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(bind_engine)

ld: error: undefined symbol: write
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(dummy_pause_job)

ld: error: undefined symbol: read
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(dummy_pause_job)

ld: error: undefined symbol: pipe
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(dummy_pause_job)

ld: error: undefined symbol: close
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(dummy_pause_job)
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(dummy_pause_job)
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(wait_cleanup)
>>> referenced by e_dasync.c
>>>               engines/dasync-dso-e_dasync.o:(wait_cleanup)
cc: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[1]: *** [Makefile:23348: engines/dasync.so] Error 1
gmake: *** [Makefile:3206: build_sw] Error 2
`

Indeed a specific setting for OpenBSD might fit better I ll try to come up with something within this weekend

**Comment by kaduk** [comment]
Could we also see the linker invocation?

**Comment by t8m** [comment]
This got updated and needs a new review

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Given this is something that would be under the current rules acceptable even to 1.1.1, can it be merged into 3.0?

**Comment by devnexen** [comment]
so long :-) I almost forgot this one

**Comment by mattcaswell** [comment]
OTC: This is too late for 3.0.0, but would be ok for 3.0.1

**Comment by t8m** [comment]
This was somehow lost.

**Comment by t8m** [comment]
Could you please rebase it against fresh master branch checkout?

**Review by hlandau — APPROVED** [review]

File: util/perl/OpenSSL/config.pm [hunk 1]
```diff
@@ -780,8 +780,10 @@ EOF
       [ 'powerpc64le-.*-.*bsd.*', { target => "BSD-ppc64le" } ],
       [ 'riscv64-.*-.*bsd.*',     { target => "BSD-riscv64" } ],
       [ 'sparc64-.*-.*bsd.*',     { target => "BSD-sparc64" } ],
+      [ 'ia64-.*-openbsd.*',      { target => "BSD-nodef-ia64" } ],
       [ 'ia64-.*-.*bsd.*',        { target => "BSD-ia64" } ],
       [ 'x86_64-.*-dragonfly.*',  { target => "BSD-x86_64" } ],
+      [ 'amd64-.*-openbsd.*',    { target => "BSD-nodef-x86_64" } ],
```
Comment (tmshort) [inline util/perl/OpenSSL/config.pm:786]: Nit: indentation?

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master. Thank you.

**Pull request closed without merging**